### ISSUE TITLE
Daemon is not working on Google Cloud Platform VM

### ISSUE THREAD

**Issue by 9orky** [newsnowlabs/docker-ingress-routing-daemon#4]
Hello @struanb
First - many thanks for your amazing work - your solution to this problem is very elegant and neat. It works perfectly on both my environments - local Vagrant and VM hosted on OVH machine. But the problem is this cannot run in GCP Compute Engine (VM).

**GCP Setup**
Our VM in GCP is available at some PUBLIC_IP. This VM is also connected to the thing called Virtual Private Cloud which is a private network which connects us with other GCP services. VM's IP in this private network is 10.156.0.4 and on this address our Swarm is listening (Swarm manager is the only machine so far, but when we will scale we'll use this private network to join other swarm nodes). Swarm has reverse proxy service (nginx) configured to be the one who listens on 80/443. 

**Problem**
Unfortunately when I start docker-ingress-routing-daemon it makes all TCP connections from PUBLIC_IP to reverse proxy to stop on SYN message - it simply tries establish a connection but no SYN-ACK is never sent back from reverse proxy. 

Do you have any idea what may cause this kind of behavior?

**Comment by struanb** [comment]
Hi @9orky. Thanks very much for your feedback. I'm glad this tool proving useful to others, and that it's working for you in various setups.

I'm happy to consider the GCP issue. We use GCP at NewsNow so I do have some experience of it and access to GCP resources for testing.

However before I would do so I'd like to be really confident I understand your setup so I could reproduce it. Could you sketch for me the minimum steps I would need to go through to setup GCP instance(s) in such as way as to reproduce the issue? 

(Alternatively if you could isolate some clean GCP instance(s) from your network and give me access to them, I could poke around, but I realise that is far easier said than done!)

**Comment by 9orky** [comment]
> However before I would do so I'd like to be really confident I understand your setup so I could reproduce it. Could you sketch for me the minimum steps I would need to go through to setup GCP instance(s) in such as way as to reproduce the issue?

@struanb Of course, here it is:

1. After creating new project we have a VPC network named "default". If you go into details of this network you will see that every Region has its own IP addresses range - in our case region is set to europe-west-3, so IPs are within 10.156.0.0/20. **Subnet creation mode** is set to automatic and **Dynamic routing mode** is set to Regional.

2. Next step was to create a VM instance (e2-standard-2) with firewall rules allowing communication on ports 80/443. We have reserved a static External IP address (not ephemeral) for the machine. VM is part of "default" network and internal IP we have obtained is 10.156.0.4

3. We have deployed our Swarm stack to the machine without IRD (Ingress Routing Daemon) first to check if installation is fine. When we made sure it works as expected we scale down our reverse proxy service and in another console we run IRD service. After that in previous console window we scale up back reverse proxy which started without any issues but the web service became unavailable - when try to run it via web browser we experience timeouts only.

4. When using nmap to check open ports it doesn't show 80/443 as open, when using TCPDUMP inside ingress_sbox it shows that SYN packets are send from PUBLIC_IP to reverse proxy service (10.0.0.8) but no reply (SYN-ACK) comes back.

5. After killing IRD process in the other window and running `ingress-routing-daemon.sh --uninstall` everything goes back to normal, but the "real IP problem" remains unsolved - it is still 10.0.0.2

**Comment by struanb** [comment]
@9orky Thanks for the detailed explanation.

One theory is that Google Cloud isn't honouring the TOS byte.

But to be clear, is the VM instance you mention in step 2, the only swarm node? i.e. You are running containers on the same node used for routing? If other swarm nodes, where are these running?

I think then I can probably launch instance(s) to try and reproduce.

**Comment by 9orky** [comment]
> But to be clear, is the VM instance you mention in step 2, the only swarm node?
@struanb Yes, so far we use only one VM with one Swarm node which is a manager-one, nothing else. As you said we run containers on the same node used for routing.

**Comment by 9orky** [comment]
`But to be clear, is the VM instance you mention in step 2, the only swarm node?`

@struanb Yes, so far we use only one VM with one Swarm node which is a manager-one, nothing else. As you said we run containers on the same node used for routing.

**Comment by struanb** [comment]
@9orky Ok then it should be straightforward for me to reproduce. If I understand correctly, I just need to launch an instance, (potentiall assign a non-ephemeral public IP), install docker, initialise a swarm, create a service, test the service, scale down, launch IRD, scale up, test the service again. I should then experience the problem (assuming it is not sensitive to the choice of internal IP/subnet).

Let me know if that sounds correct, and I'll try and make time this weekend.

P.S. Regarding the internal subnet, just a thought: could there be some conflict between the internal instance subnet and the docker ingress subnet? If so, that could potentially affect routing.

**Comment by 9orky** [comment]
> Ok then it should be straightforward for me to reproduce. If I understand correctly, I just need to launch an instance, (potentiall assign a non-ephemeral public IP), install docker, initialise a swarm, create a service, test the service, scale down, launch IRD, scale up, test the service again. I should then experience the problem (assuming it is not sensitive to the choice of internal IP/subnet).

You are correct. In our setup this service is Nginx (as a reverse proxy) and for the sake of compatibility with our setup let there be another Nginx service which is an upstream defined in reverse proxy. Not sure if it makes a difference but this is how it looks in our case :)

> P.S. Regarding the internal subnet, just a thought: could there be some conflict between the internal instance subnet and the docker ingress subnet? If so, that could potentially affect routing.

This is what I've been thinking too, that's why I have created another network with 192.168.0.0/24 range and problem was the same. Both networks (10.0.0.0/24 and 192.168.0.0/24) are configured with default values, no routing rules or extra firewall rules were applied by me. If there is a conflict caused by the IP resolution I have no idea what it would be :( but without IRD running it works well, so maybe there is something within the GCP infrastructure what manipulates packets in some unexpected way?

**Comment by struanb** [comment]
Hi @9orky. Thanks to your detailed instructions it only took me 15 minutes to reproduce your issue this afternoon. Tracking down the cause took me rather longer!

I believe the issue, at least with the `4.19.0-14-cloud-amd64` kernel that came with Google's Debian Buster image, is that the rp_filter sysctl value is being set to 1 by default on all interfaces (and therefore on the ingress network interface) inside all launched containers. This prevents reverse path packets from being correctly processed by the networking stack.

Setting it to 2 - loose mode - appears to be sufficient to resolve the issue.

The following patch worked for me. Could you try it?

```
--- docker-ingress-routing-daemon-3.0.1 2021-03-06 15:15:26.439545481 +0000
+++ docker-ingress-routing-daemon-3.1   2021-03-06 15:17:33.742370103 +0000
@@ -279,6 +279,10 @@
     # 3. Map any connection mark on outgoing traffic to a firewall mark on the individual packets.
     nsenter -n -t $NID iptables -t mangle -A OUTPUT -p tcp -j CONNMARK --restore-mark
 
+    # 3.1 Enable 'loose' rp_filter mode on interface $CIF (and 'all' as required by kernel
+    #     see https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt)
+    nsenter -n -t $NID sysctl net.ipv4.conf.all.rp_filter=2 net.ipv4.conf.$CIF.rp_filter=2
+
     for NODE_IP in $INGRESS_NODE_GATEWAY_IPS
     do
       NODE_ID=$(echo $NODE_IP | cut -d'.' -f4)
```

**Comment by 9orky** [comment]
@struanb Yes, it worked like a charm! I can't find words to describe how grateful we are for your solution and for the time you have spent solving our problem. Is there any way we could repay you? Wish we could have a beer or two someday :) 

As for me that's it in this subject - think we can safely close this issue.

**Comment by struanb** [comment]
@9orky That's great news and you're welcome! Beer in a post-lockdown world would be nice, but actually I need to thank you for trying the IRD out and reporting this problem (which is good to have solved, not least as I expect we would have hit it ourselves in production sooner or later when we upgraded our raw metal kernels).

Please do report any further issues. And if you can, please spread the word about IRD, as there seems to have been many people who've experienced this docker service limitation over many years.

And please look out the next projects we'll be open-sourcing.

**Comment by struanb** [comment]
This fix is now released in v3.1.0.
